packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp

```diff
@@ -5927,42 +5927,36 @@ namespace Tpetra {
         Distributor& dist) const
   {
     using Details::packCrsMatrix;
-
     if (this->isStaticGraph ()) {
-      const map_type& colMap = * (this->staticGraph_->colMap_);
-      const auto lclColMap = colMap.getLocalMap ();
-      const int myRank =
-        colMap.getComm ().is_null () ? 0 : colMap.getComm ()->getRank ();
-      std::unique_ptr<std::string> errStr;
+      const Teuchos::Comm<int>& comm = *(this->getComm());
+      const int myRank = comm.getRank();
+      auto pack_result =
+        packCrsMatrix (*this, exportLIDs, exports, numPacketsPerLID,
+                       constantNumPackets, dist);
+      const int lclOK = pack_result.success() ? 1 : 0;
 #ifdef HAVE_TPETRA_DEBUG
       using Teuchos::outArg;
       using Teuchos::REDUCE_MIN;
       using Teuchos::reduceAll;
-      const bool locallyCorrect =
-        packCrsMatrix (this->lclMatrix_, lclColMap, errStr,
-                       exports, numPacketsPerLID, constantNumPackets,
-                       exportLIDs, myRank, dist);
-      const int lclOK = locallyCorrect ? 1 : 0;
       int gblOK = 1; // output argument
-      if (! colMap.getComm ().is_null ()) {
-        reduceAll<int, int> (* (colMap.getComm ()), REDUCE_MIN,
-                             lclOK, outArg (gblOK));
-        if (gblOK != 1) {
-          std::ostringstream out;
-          if (colMap.getComm ()->getRank () == 0) {
-            out << "Error in packCrsMatrix!" << std::endl;
-          }
-          using ::Tpetra::Details::gathervPrint;
-          const std::string errStr2 =
-            errStr.get () == NULL ? std::string ("") : *errStr;
-          gathervPrint (out, errStr2, * (colMap.getComm ()));
-          TEUCHOS_TEST_FOR_EXCEPTION(true, std::runtime_error, out.str ());
-        }
+      reduceAll<int, int> (comm, REDUCE_MIN, lclOK, outArg (gblOK));
+      if (gblOK != 1) {
+        using ::Tpetra::Details::gathervPrint;
+        std::ostringstream out_v, out;
+        out << "ERROR: packCrsMatrix returned the following "
+          << "error on Proc " << myRank << ":\n";
+        const std::string errStr = out.str() + pack_result.get_error_str();
+        gathervPrint (out_v, errStr, comm);
+        TEUCHOS_TEST_FOR_EXCEPTION(true, std::runtime_error, out_v.str ());
       }
 #else // NOT HAVE_TPETRA_DEBUG
-      (void) packCrsMatrix (this->lclMatrix_, lclColMap, errStr,
-                            exports, numPacketsPerLID, constantNumPackets,
-                            exportLIDs, myRank, dist);
+      if (!lclOK) {
+        std::ostringstream out;
+        out << "ERROR: packCrsMatrix returned the following "
+            << "error on Proc " << myRank << ":\n";
+        const std::string errStr = out.str() + pack_result.get_error_str();
+        TEUCHOS_TEST_FOR_EXCEPTION(true, std::runtime_error, errStr);
```

Comment:
If you use the `TEUCHOS_TEST_FOR_EXCEPTION*` macros, prefer `TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC`, since it prints the class' name as part of the message.  Use of the latter macro requires that the method define the `tfecfFuncName` string.  See examples in `Tpetra_CrsMatrix_def.hpp`.